### ISSUE TITLE
NPE triggered by liquibase.Scope.getLog

### ISSUE THREAD

**Issue by Treehopper** [liquibase/liquibase#1354]
## Environment
**Liquibase Version**: 4.0.0

## Description
I am getting an NPE when I create a new FileSystemResourceAccessor(rootPath), with the following stack trace:

```
java.lang.NullPointerException
        at liquibase.Scope.getLog(Scope.java:278)
        at liquibase.resource.FileSystemResourceAccessor.addRootPath(FileSystemResourceAccessor.java:54)
        at liquibase.resource.FileSystemResourceAccessor.<init>(FileSystemResourceAccessor.java:36)
```

Looking at the Liquibase code, I see that SmartMap#get defines its result as a nullable, however, the Scope class does not check for null in this and other places.

## Steps To Reproduce
I'm running tests in my project, using liquibase-core:
Created this issue for tracking purposes, where you can also find the full stack trace: https://github.com/Treehopper/liquibase-lsp/issues/17

## Additional Context
In my project, I have to run scripts asynchronously. This, and the fact that the problem only occurs occasionally hints at a race-condition.

**Comment by molivasdat** [comment]
Hi @Treehopper .  Thank you for writing up this issue.  What databases are you connecting to when this error occurs?  You mentioned that you are running tests in your project.  What tests are you running?  We definitely need to make sure we don't have an NPE.  Do you have a suggested fix for this and a PR to go along with that?

**Comment by Treehopper** [comment]
Hi @molivasdat . I am connecting to an H2 embedded DB. The tests are JUnit tests, integrating several components in addition to Liquibase.
I agree that an NPE should not occur, but I believe how a potential fix would look like, depends on the question whether liqubase-core should be thread-safe or not, or if the user of the liquibase-core API has to ensure that no concurrent access occurs?
The fact that I find 47 'synchronized' statements seem to indicate that liquibase-core should actually be thread-safe, and as a user of its API, I don't have to worry about accessing it in concurrent threads.
Can you confirm this assumption about the design?

Some further analysis:
I believe the root-cause is to be found in the Scope-switching mechanic:
The Scope class ensures via its private constructor that there is a default LogService, and the only way to overwrite it is via a protected constructor. If that were used anywhere to set a null LogService, the `get` method would still check for it and look for an alternative in the parent. This is a recursive call, so it should eventually find that default LogService in the rootScope. This seems fail-safe.
However, it is possible to call the protected constructor with a null-parent (no check here). If that would happen, that scope would be detached, and the `get` method may return null.
How can a null-parent be set?
In `Scope.child(LiquibaseListener, Map<String, Object>, ScopedRunnerWithReturn<T>)` `getCurrentScope()` is used to retrieve/create&set a Scope via the static `scopeManager`. Access on the `scopeManager `is not synchronized.
Thus, overlapping calls to `getCurrentScope()` may result in different threads overriding the static `scopeManager` they share, and the `rootScope` being set in the "wrong" `scopeManager`, leaving the `scopeManager` with a null `currentScope`.
Another call to `getCurrentScope()` would then return a null Scope, because only the manager is checked for null.
If that call came from the aforementioned `child` method, we get a child with a null-parent, resulting in a null-value returned from the `get`.
There are two next steps I can think of:
- Attempt to write a test to reproduce the described behavior reliably (this might be a tough one)
- Add a null-check for the parent to the protected constructor, and throw an `IllegalArgumentException` (failing early is always good).

Apart from the NPE, I also found an occassional `ConcurrentModificationException`. I have not created a Liquibase ticket yet, because I wanted to wait for a response on this ticket, but you can find the stack trace here:
https://github.com/Treehopper/liquibase-lsp/issues/1
This raises the same question, though: Should liquibase-core be thread-safe?

**Comment by nvoxland** [comment]
Liquibase-core SHOULD be thread-safe. However, the integration points we have are all single-threaded so it's not tested as well as it should be. And, like you say, writing test cases for the multi-threading race conditions is tough....

I think a check for a not-null parent in the protected Scope constructor makes sense. If you could do a PR for that, that would be great.

And, as you find other concurrency issues, tickets/PRs for them would be good too.

The ScopeManager is designed at some point to be pluggable, so we could have ThreadLocalScopeManager (where each thread has a separate Scope) etc. That has not been implemented, but something to be thinking about.

**Comment by Treehopper** [comment]
> And, as you find other concurrency issues, tickets/PRs for them would be good too.

I did randomly find two more NPEs in SmartMap and in SqlGeneratorFactory.
Let's merge the above PR for the constructor check before raising new tickets, though.

**Comment by Treehopper** [comment]
Seems like one of the issues I have encountered has been fixed in the scope of this ticket: https://github.com/liquibase/liquibase/issues/1402

**Comment by nvoxland** [comment]
I believe the issue this is describing (NPE triggered by liquibase.Scope.getLog) was handled in #1402. The referenced PR #1377 is a good generic fix, but unrelated to this specific issue.

Is that right, @Treehopper?

**Comment by Treehopper** [comment]
@nvoxland I agree that #1377 was a general defensive-programming-improvement. Thanks for updating.
With respect to this issue, however, it was more than that, as it should prove the theory I have explained in the initial description, but wasn't 100% confident about.
As I don't know the code very well, and the error I saw was only a follow-up problem of this assumed race-condition, I wanted to provide this proof as a justification for another change.
If the exception would be thrown in the described scenarion, after merging the change, the next step might be for example synchronizing access on the scopeManager.
As I mentioned in [this ](https://github.com/liquibase/liquibase/issues/1354#issuecomment-686417374) comment, similar changes would be required elsewhere.
To anybody using Liquibase in a multi-threaded environment, I'd recommend to synchronize access in the code calling the API, until these are fixed.

I'm not actively working on the project anymore which would use Liquibase in this way, so I cannot give more practical feedback, and probably won't be able to provide that follow-up fix anytime soon, though.
